### PR TITLE
Include Tax on enterprise fees on the invoice

### LINKED ISSUE #12021: Tax on enterprise fees is not displaying on the invoice
## Description
If an enterprise fee incurs tax,  it is not displaying on the invoice.   The tax on the fee IS correctly calculating,  and the order tax total is correct,  but the line item for the product is missing the **tax on the fee.**

This only seems to be a problem for fees added to the product (per item fees).   Tax is correctly calculating on whole order fees. 

## Expected Behavior
On   invoice table,  the 'sales tax' column should show the tax on the product + the tax on any fees on the product.

## Actual Behaviour
On the invoice table,  the 'sales tax' column is showing the tax on the product and NOT the tax on the product fee.

## Steps to Reproduce
1.Create a product with a per product enterprise fee
2.Ensure both the product,  and the enterprise fee have a tax category assigned
3.Place an order (note the order confirmation and total correctly calculates product cost + product tax + product fee + product fee tax)
4.See the invoice (current and new) 

Current and New invoices with annotation below
invoice-R642522773 pdf_page-0001
line fee tax incorrect on new invoice _page-0001

See also issue - https://github.com/openfoodfoundation/openfoodnetwork/issues/12066

## Workaround
No workaround -  invoices are displaying incorrect information -  quite confusing to users.

## Severity
s2

## Your Environment

* Version used:
* Browser name and version:
* Operating System and version (desktop or mobile):

### PR DESCRIPTION
**Pull request by rioug**
#### What? Why?

- Closes #12021 

In the case of tax not included in the price, when a Enterprise Fee incurs a tax, currently it's not included in the line item tax column on the invoice. This fix the issue, by adding the tax portion of the fee to the line item tax.
It also fix a range of issue where the enterprise fee tax was wrongly added to some total, or missing from the tax total, see testing notes for more information.

NOTE TO REVIEWERS:  the git history could be cleaner but I already spent too long on this. I can clean it up if it necessary.

#### Tax excluded from price

In the backend:
- Make sure you have a tax rate setup with tax excluded from price, see Configuration -> Tax rates
- Set up an Enterprise fee for your testing enterprise, see Configuration -> Enterprise Fees or Enterprise Fees from the enterprise setting page
- Create a product with a per product enterprise fee
- Ensure both the product, and the enterprise fee have a tax category assigned

##### With old invoice (invoices feature disable)

- Place an order with said produce
- Go to the order in the back office, and click on actions -> Print Invoice
- Check the GST (could also be named "Sales tax" depending on your configuration) include the Enterprise Fee tax. Enterprise fee tax are not shown in the backoffice, but you should be able to calculate/infer the value from the Enterprise Fee line item adjustment and your configuration

old_invoice

##### With new invoice (invoices feature enable)
- As above, place an order and click print invoice
- Check the Line Item Total Price (Incl. Tax) includes the Enterprise Fee Tax

new_invoice

#### Tax included in price
Use the setup described above but switch to price included in tax.

##### With old invoice (invoices feature disable)

- Place an order with said produce
- Go to the order in the back office, and click on actions -> Print Invoice
- Check the GST (could also be named "Sales tax" depending on your configuration) include the Enterprise Fee tax. Enterprise fee tax are not shown in the backoffice, but you should be able to calculate/infer the value from the Enterprise Fee line item adjustment and your configuration

old_invoice_incl_tax

##### With new invoice (invoices feature enable)
- As above, place an order and click print invoice
- Check the Price Per Unit ( Excl. Tax ) does not include the enterprise fee tax
- Check the Total ( Excl. Tax ) does not include the enterprise fee tax

new_invoice_incl_tax

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [x] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [ ] Technical changes only
- [ ] Feature toggled

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: app/models/enterprise_fee_adjustments.rb
Comment (rioug) [inline app/models/enterprise_fee_adjustments.rb:1]: This could also live in the `LineItem` model, but I didn't not think it was the responsibility of the `LineItem` to know about the enterprise fee tax. So I added this small model to take care of it, and it makes testing easier.

**Review by mkllnk — APPROVED** [review]
Just scanned through but it all looks reasonable.

**Review by dacook — COMMENTED** [review]
Great, thanks so much for working through this. Handling different tax options is so complicated!

I think this is fine as it is, but after thinking about it I have an idea to simplify (and I'm keen because I find the tax logic hard to follow already!)

What do you think of the new methods being in a module/concern, included on LineItem? _(edited to make more sense)_
This means the methods can be called directly on LineItem and allows the tax_helper and serialisers to stay nice and simple.
The tests can continue to live in a separate file like the [concern tests](https://github.com/openfoodfoundation/openfoodnetwork/tree/master/spec/models/concerns) we currently have.

As this is an S2, I'm not sure if we should wait to do this, or just proceed for a faster response.. I'll ask in  #delivery-circle.

**Review by dacook — APPROVED** [review]
I'm approving in case this needs to move forward quicker, but if it can wait I suggest the refactor (or at least the chance to discuss when we're both online next week).

File: app/models/invoice/data_presenter/line_item.rb [hunk 1]
```diff
@@ -30,9 +32,11 @@ def display_amount_with_adjustments_and_with_taxes
       end
 
       def single_display_amount_with_adjustments
-        Spree::Money.new(price_with_adjustments - (included_tax / quantity), currency:)
+        fee_tax = enterprise_fee_included_tax || 0.0
+        Spree::Money.new(price_with_adjustments - ((included_tax + fee_tax) / quantity), currency:)
       end
 
+      # TODO seems useless
```
[the remark this answers is not in the document]
Comment (rioug) [reply app/models/invoice/data_presenter/line_item.rb]: I meant to remove that, thanks !

File: app/models/enterprise_fee_adjustments.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+# EnterpriseFeeAdjustments represent a collection of enterprise fee adjustments
+#
+class EnterpriseFeeAdjustments
+  def initialize(adjustments)
+    @adjustments = adjustments
+  end
+
+  # Calculate the tax portion of enterprise fee when tax excluded from price
+  def total_tax
+    @adjustments.reduce(0.0) do |sum, enterprise_fee|
```
[the remark this answers is not in the document]
Comment (rioug) [reply app/models/enterprise_fee_adjustments.rb:22]: Indeed you could use `sum`, it makes to code a bit cleaner. That prompted me to improve the specs as well.

**Review by mkllnk — APPROVED** [review]

File: app/helpers/tax_helper.rb [hunk 3]
```diff
@@ -12,12 +12,15 @@ def display_taxes(taxable, display_zero: true)
     end
   end
 
-  def display_line_items_taxes(line_item, display_zero: true)
+  def display_line_items_taxes(
+    line_item, display_zero: true, enterprise_fee_adjustments: EnterpriseFeeAdjustments.new
+  )
+    enterprise_fee_adjustments.adjustments = line_item.enterprise_fee_adjustments
```
Comment (dacook) [inline app/helpers/tax_helper.rb:18]: I don't see how this is any less coupled, we're still instantiating `EnterpriseFeeAdjustments` in the same class, just in a different (I think more convoluted) way.

I have to admit I didn't know the "dependency injection" concept though. I looked it up and can see how it could be achieved by passing an instantiated object, but I don't see that happening here. I'm not sure if it's necessary.

**Review by dacook — CHANGES_REQUESTED** [review]
Thanks for following up, but I have to admit I don't follow, and it seems to make things more complicated for no benefit. So I'd suggest rewinding the last two commits, but maybe I'm not understanding?

**Review by dacook — APPROVED** [review]
I've discussed with Gaetan, and agreed to rewind to the previous commit which was approved, so we can move this forward.

**Comment by drummer83** [comment]
Hi @rioug,
Thanks for fixing this.

I have tested the following before and after staging your PR:
- Check invoices with tax excl. from price for enterprise fee per **item**
- Check invoices with tax incl. in price for enterprise fee per **item**

I also tested this after staging only:
- Check invoices with tax excl. from price for enterprise fee per **order**
- Check invoices with tax incl. in price for enterprise fee per **order**

The following abbreviations are used in the table: STD = current standard invoice template; ALT = current alternative invoice template; NEW = new unified invoice template.

Enterprise fee per item:
| Tax | Before STD | After STD | Before ALT | After ALT | Before NEW | After NEW |
| - | - | - | - | - | - | - |
| Excl. | :x: | :heavy_check_mark: :muscle: | :heavy_check_mark: | :heavy_check_mark: | :x: | :x: (see below) |
| Incl. | :x: | :heavy_check_mark: :partying_face: | :heavy_check_mark: | :heavy_check_mark: | :x: | :x: (see below) |

Enterprise fee per order:
| Tax | Before STD | After STD | Before ALT | After ALT | Before NEW | After NEW |
| - | - | - | - | - | - | - |
| Excl. |  | :heavy_check_mark: |  | :heavy_check_mark: |  | :heavy_check_mark: |
| Incl. |  | :heavy_check_mark: |  | :heavy_check_mark: |  | :heavy_check_mark: |

## Remaining issues
Using the new invoice template we still have incorrect numbers for both tax incl. and tax excl. Here are the examples:

Tax excl. from price:
image

Tax incl. in price:
image

## Conclusion
The still incorrect numbers are behind a feature toggle, so I think it is safe to merge this one and open an issue for the remaining bits.

Thanks Gaetan for all your work on this. :pray: 
Merging! :rocket:

**Comment by drummer83** [comment]
New issue created here #12106.
